src/style-spec/function/expression.js

```diff
@@ -123,8 +125,87 @@ class ParsingContext {
     }
 }
 
+class CompilationContext {
+    _id: number;
+    _cache: {[string]: string};
+    _prelude: string;
+    scope: Scope;
+
+    constructor() {
+        this._cache = {};
+        this._id = 0;
+        this._prelude = '';
+        this.scope = new Scope();
+    }
+
+    compile(e: Expression): string {
+        const id = this.addExpression(e.compile(this));
+        return `${id}()`;
+    }
+
+    compileToFunction(e: Expression, evaluationContext: Object): Function {
+        const finalId = this.addExpression(e.compile(this));
+        const src = `
+var $globalProperties;
+var $feature;
+var $props;
+${this._prelude}
+return function (globalProperties, feature) {
+    $globalProperties = globalProperties;
+    $feature = feature;
+    $props = feature && $feature.properties || {};
+    return $this.unwrap(${finalId}())
+};`;
+        return (new Function('$this', src): any)(evaluationContext);
+    }
+
+    getPrelude() {
+        return this._prelude;
+    }
+
+    addExpression(body: string): string {
+        let id = this._cache[body];
+        if (!id) {
+            id = `e${this._id++}`;
+            this._cache[body] = id;
+
+            // convenience: allow simple expressions to omit 'return ' in
+            // compiled string
+            if (!/return/.test(body)) {
+                body = `return ${body}`;
+            }
```

Comment:
I think it might be better to write out `return` explicitly than introduce non-obvious shortcuts like this...

Reply:
Agreed.  I never did feel good about this 😄

Reply:
@mourner I'm going back on this.  An advantage to having something like this is that, when it's possible for an expression to be compiled to a pure JS expression that can be easily inlined (rather than a function body, which would be more complicated to inline), we could do so.  I know you said that JS engines will inline pure functions w/o side effects, but I'd still like to leave this open for further investigation.

Reply:
(That said, the current regex check is definitely not the way to do this.)

Reply:
How about converting all keywords to be pure expressions and then always adding a `return` here? This way all expression would potentially be inlinable and no need for this conditional logic. I went through all expressions in this ticket and it looks like they all could be converted.

Reply:
I'm not sure they can all be pure expressions, _unless_ we commit to providing a helper method in `EvaluationContext` for any case where we need to use a subexpression's value more than once.